File: 823-binary-trees-with-factors/823-binary-trees-with-factors.js
var numFactoredBinaryTrees = function(A) {
    A.sort((a,b) => a - b)
    let len = A.length, fmap = new Map(), ans = 0
    for (let i = 0; i < len; i++) {
        let num = A[i], ways = 1, lim = Math.sqrt(num)
        for (let j = 0, fA = A[0]; fA <= lim; fA = A[++j]) {
            let fB = num / fA
            if (fmap.has(fB))
                ways += fmap.get(fA) * fmap.get(fB) * (fA === fB ? 1 : 2)
        }
        fmap.set(num, ways), ans += ways
    }
    return ans % 1000000007
};

File: 823-binary-trees-with-factors/823-binary-trees-with-factors.java
class Solution {
    public int numFactoredBinaryTrees(int[] A) {
        Arrays.sort(A);
        int len = A.length;
        long ans = 0;
        HashMap<Integer, Long> fmap = new HashMap<>();
        for (int num : A) {
            long ways = 1;
            double lim = Math.sqrt(num);
            for (int j = 0, fA = A[0]; fA <= lim; fA = A[++j]) {
                if (num % fA != 0) continue;
                int fB = num / fA;
                if (fmap.containsKey(fB))
                    ways += fmap.get(fA) * fmap.get(fB) * (fA == fB ? 1 : 2);
            }
            fmap.put(num, ways);
            ans = (ans + ways) % 1000000007;
        }
        return (int)ans;
    }
}